File: src/io/innofang/prototype/example/WordDocument.java
package io.innofang.prototype.example;

import java.util.ArrayList;

/**
 * Created by Inno Fang on 2017/2/27.
 */
public class WordDocument implements Cloneable {

    private String text;
    private ArrayList<String> images = new ArrayList<>();

    public WordDocument() {
        System.out.println("-------------init-------------");
    }


    public String getmText() {
        return text;
    }

    public void setText(String mText) {
        this.text = mText;
    }

    public ArrayList<String> getmImages() {
        return images;
    }

    public void setImages(ArrayList<String> mImages) {
        this.images = mImages;
    }

    public void addImage(String image) {
        images.add(image);
    }

    public void showDocument() {
        System.out.println("-------------Start-------------");
        System.out.println("Text : " + text);
        System.out.println("Image List : ");
        for (String mImage : images) {
            System.out.println("Image Name : " + mImage);
        }
        System.out.println("------------- End -------------");
    }

    protected WordDocument clone() {
        try {
            WordDocument copy = (WordDocument) super.clone();
            copy.text = this.text;
//            copy.images = this.images;
            copy.images = (ArrayList<String>) this.images.clone();
            return copy;
        } catch (CloneNotSupportedException e) {
            e.printStackTrace();
        }
        return null;
    }

}

File: src/io/innofang/prototype/example/kotlin/WordDocument.kt
package io.innofang.prototype.example.kotlin

import java.util.ArrayList

/**
 * Created by Inno Fang on 2017/8/30.
 */
class WordDocument
constructor(var text: String = "", var images: ArrayList<String> = ArrayList<String>()) : Cloneable {

    init {
        println("----- init -----")
    }

    fun addImage(image: String) {
        images.add(image)
    }

    fun showDocument() {
        println("---- start -----")
        println("Text: " + text)
        println("Image List : ")
        images.map { println("Image name : $it") }
        println("----- End ------")
    }

    fun cloneTo(): WordDocument? {
        try {
            val copy: WordDocument = super.clone() as WordDocument
            copy.text = this.text
            copy.images = this.images.clone() as ArrayList<String>
            return copy
        } catch (e: CloneNotSupportedException) {
            e.printStackTrace()
        }
        return null
    }
}